substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrTypeRepository.java

```diff
@@ -54,26 +54,30 @@ public long getClassId(Class<?> clazz) {
     }
 
     @Override
-    public int write(JfrChunkWriter writer) {
+    public int write(JfrChunkWriter writer, boolean flush) {
         // Visit all used classes, and collect their packages, modules, classloaders and possibly
         // referenced classes.
-        TypeInfo typeInfo = collectTypeInfo();
+        TypeInfo typeInfo = collectTypeInfo(flush);
 
         // The order of writing matters as following types can be tagged during the write process
-        int count = writeClasses(writer, typeInfo);
-        count += writePackages(writer, typeInfo);
-        count += writeModules(writer, typeInfo);
-        count += writeClassLoaders(writer, typeInfo);
+        int count = writeClasses(writer, typeInfo, flush);
+        count += writePackages(writer, typeInfo, flush);
+        count += writeModules(writer, typeInfo, flush);
+        count += writeClassLoaders(writer, typeInfo, flush);
         count += writeGCCauses(writer);
         count += writeGCNames(writer);
         count += writeVMOperations(writer);
         return count;
     }
 
-    private TypeInfo collectTypeInfo() {
+    private TypeInfo collectTypeInfo(boolean flush) {
         TypeInfo typeInfo = new TypeInfo();
         for (Class<?> clazz : Heap.getHeap().getLoadedClasses()) {
-            if (JfrTraceId.isUsedPreviousEpoch(clazz)) {
+            if (flush) {
+                if (JfrTraceId.isUsedCurrentEpoch(clazz)) {
+                    visitClass(typeInfo, clazz);
```

Comment:
I think the reason I haven't seen any issues with the current method is because the constant pools are fully flushed every time. I think the non-constant IDs will become an issue once only new constant data is flushed. I'll update this.

Reply:
Do you think a reasonable approach would be to use normal (interruptible) maps to record the IDs of packages, modules, and classloaders? This would only ever be used when the chunk writer lock is acquired during a flush or an epoch change. The maps would be reset after an epoch change.

Reply:
Interruptible maps should work in this particular case.